$OpenBSD: patch-lib_puppet_provider_user_useradd_rb,v 1.4 2014/11/04 09:03:58 jasper Exp $

- add new 'loginclass' attribute to the 'user' type
- make 'chage' an optional command

--- lib/puppet/provider/user/useradd.rb.orig	Tue Oct 21 23:52:27 2014
+++ lib/puppet/provider/user/useradd.rb	Sun Nov  2 20:55:27 2014
@@ -9,7 +9,7 @@ Puppet::Type.type(:user).provide :useradd, :parent => 
     install Ruby's shadow password library (often known as `ruby-libshadow`)
     if you wish to manage user passwords."
 
-  commands :add => "useradd", :delete => "userdel", :modify => "usermod", :password => "chage"
+  commands :add => "useradd", :delete => "userdel", :modify => "usermod"
 
   options :home, :flag => "-d", :method => :dir
   options :comment, :method => :gecos
@@ -17,6 +17,7 @@ Puppet::Type.type(:user).provide :useradd, :parent => 
   options :password_min_age, :flag => "-m", :method => :sp_min
   options :password_max_age, :flag => "-M", :method => :sp_max
   options :password, :method => :sp_pwdp
+  options :loginclass, :flag => '-L', :method => :sp_loginclass
   options :expiry, :method => :sp_expire,
     :munge => proc { |value|
       if value == :absent
@@ -42,6 +43,8 @@ Puppet::Type.type(:user).provide :useradd, :parent => 
     }
 
   optional_commands :localadd => "luseradd"
+  optional_commands :chage    => "chage"
+  optional_commands :password => "passwd"
   has_feature :libuser if Puppet.features.libuser?
 
   def exists?
@@ -159,6 +162,7 @@ Puppet::Type.type(:user).provide :useradd, :parent => 
       next if property.to_s =~ /password_.+_age/
       next if property == :groups and @resource.forcelocal?
       next if property == :expiry and @resource.forcelocal?
+      next if property == :loginclass and @resource.forcelocal?
       # the value needs to be quoted, mostly because -c might
       # have spaces in it
       if value = @resource.should(property) and value != ""
@@ -191,16 +195,25 @@ Puppet::Type.type(:user).provide :useradd, :parent => 
     cmd << @resource[:name]
   end
 
+  def self.passwordcmd
+    case Facter.value(:operatingsystem)
+    when 'OpenBSD'
+      command(:password)
+    else
+      command(:chage)
+    end
+  end
+
   def passcmd
     age_limits = [:password_min_age, :password_max_age].select { |property| @resource.should(property) }
     if age_limits.empty?
       nil
     else
-      [command(:password),age_limits.collect { |property| [flag(property), @resource.should(property)]}, @resource[:name]].flatten
+      [passwordcmd,age_limits.collect { |property| [flag(property), @resource.should(property)]}, @resource[:name]].flatten
     end
   end
 
-  [:expiry, :password_min_age, :password_max_age, :password].each do |shadow_property|
+  [:expiry, :password_min_age, :password_max_age, :password, :loginclass].each do |shadow_property|
     define_method(shadow_property) do
       if Puppet.features.libshadow?
         if ent = Shadow::Passwd.getspnam(@resource.name)
